Introduce GPG config file changes to accomodate gpg 2.1

libBletchley opened the pull request:
* **What does this PR do?**
 * renames `gpg.rc` to `gpg2.rc`
 * introduces `gpg2_1.rc` which brings in a parameter that does not exist in older GPG versions
 * places gpg commands at the end of the commandline, as required by the (often unenforced) syntax rules
 * remove use of `--always-trust` in the new version (on the advice of GPG folks)
 * ultimately fixes bug https://github.com/neomutt/neomutt/issues/944

* **Are there points in the code the reviewer needs to double check?**
I didn't do any testing on the removal of `--always-trust`.

* **Does this PR meet the acceptance criteria?**
I think so.  I do not believe this change requires any of the following.

   - Documentation created/updated (you have to edit
     [doc/manual.xml.head](https://www.github.com/neomutt/neomutt/blob/master/doc/manual.xml.head)
     for that)

   - All builds are passing

   - Added [doxygen code documentation](https://www.neomutt.org/dev/doxygen)
     [syntax](http://www.stack.nl/~dimitri/doxygen/manual/docblocks.html)

   - Code follows the [style guide](https://www.neomutt.org/dev/coding-style)

* **What are the relevant issue numbers?**
https://github.com/neomutt/neomutt/issues/944

gahr commented:
Given that 2.1.0 is 3 years old, do you think it would make sense to just upgrade the file to be 2.1-compliant and drop support for < 2.1?

libBletchley commented:
Debian is always a bit behind.  Debian _Stretch_ was released ~5 months ago.  The predecessor (Jessie) uses the pre-2.1 config.  There have been moments that I was 2 Debian releases behind due to stuff breaking in next release.  I'm not sure how common that is - probably rare.

The end of life date for Jessie long-term-support is ~June 6th 2020 (https://wiki.debian.org/DebianReleases).  I'm not sure if that's a factor.. If someone is stuck on Jessie that long, they should in principle also be using the old version of mutt 1.5.23-3.  But if such a user would take a bleeding edge version of mutt it'll break no matter what, but having the old config file there would hopefully provide a quick fix.

It's your call.. I just noticed there are multiple versions of `pgp*.rc` files and followed the pattern.

libBletchley commented:
Another option would be to put the contents of `gpg2.rc` in `gpg2_1.rc`, as comments, and rename back to `gpg.rc`.

gahr commented:
Debian has a record for maintaining their own patched versions of most software. I don't think we should wait on them to drop support for EOL'd 3rd party software. Debian is free to maintain patches against gpg.rc as well.

libBletchley commented:
I've consolidated back to `gpg.rc`.  The history should be preserved the way that I did it (`git mv gpg2.rc gpg.rc`, `cp gpg2_1.rc gpg.rc`, `git rm gpg2_1.rc`).

I also noticed the `default.auto` file was a Tcl file that appears in fundamental mode in emacs due to the non-standard extension, so I added the tag to fix that.

gahr commented:
Thanks @libBletchley. A couple of notes: 

* `autosetup/default.auto` is imported from upstreams, we do not maintain that, and it has nothing to do with this issue. Please revert that change.
* I'd rather not say `Command formats for GnuPG 2.1`. If you want to specify a version, please use `>=`. I'm running 2.2.2 here.
* Could you please try to keep the diffs to a minimum? There are a few things that do not belong to this issue. Ideally, we should only see changes in `set *_command = ....` lines, right?

libBletchley commented:
> I'd rather not say Command formats for GnuPG 2.1. If you want to specify a version, please use >=. I'm running 2.2.2 here.

I've removed the version number, as I'd rather not predict what to expect of future versions.

> Could you please try to keep the diffs to a minimum? There are a few things that do not belong to this issue. Ideally, we should only see changes in set *_command = .... lines, right?

I'm not quite sure what you're referring to.  There was some noise b/c I had two config files and went back to one, but if you just diff prior to the first commit with the last commit that noise won't appear.  Or if you mean my removal of "PGP/MIME" in some of the comments, those were misleading because they were commenting commands that are also used for PGP/inline operations.  So I removed the PGP/MIME comment in the cases that were clearly both pgp/mime and inline.  I'm not sure if I got them all though.

gahr commented on contrib/gpg.rc:
```diff
@@ -1,10 +1,18 @@
 #   http://70t.de/download/gpg-2comp.tar.gz
 #
+# $Id: gpg.rc,v 1.12 2001/12/11 09:33:57 roessler Exp $
+#
```
Where does this come from?

libBletchley replied:
I forgot that I did that.  Those are anchors.  Some projects want comments to appear above the code they refer to, and some want comments to appear below the referenced code.  I've also worked on a project that had it both ways, depending on what kind of code it was.  The purpose of the anchors are to clearly attach the comment to the SLOC it refers to.

gahr commented on contrib/gpg.rc:
```diff
@@ -29,58 +37,72 @@
 # set pgp_encrypt_only_command="/usr/lib/neomutt/pgpewrap gpg-2comp -v --batch --output - --encrypt --textmode --armor --always-trust -- -r %r -- %f"
-set pgp_encrypt_only_command="/usr/lib/neomutt/pgpewrap gpg --batch --quiet --no-verbose --output - --encrypt --textmode --armor --always-trust -- -r %r -- %f"
+#
+set pgp_encrypt_only_command="/usr/lib/mutt/pgpewrap gpg --batch --quiet --no-verbose --output - --textmode --armor -- --recipient %r -- --encrypt %f"
```
please keep `/usr/lib/neomutt/`

libBletchley replied:
Ah, I overlooked that.  Debian renames neomutt to mutt, so that was Debian stuff creeping in.  I'll fix it.

gahr commented on contrib/gpg.rc:
```diff
@@ -29,58 +37,72 @@
 # set pgp_encrypt_sign_command="/usr/lib/neomutt/pgpewrap gpg-2comp %?p?--passphrase-fd 0? -v --batch --output - --encrypt --sign %?a?-u %a? --armor --always-trust -- -r %r -- %f"
-set pgp_encrypt_sign_command="/usr/lib/neomutt/pgpewrap gpg %?p?--passphrase-fd 0? --batch --quiet --no-verbose --textmode --output - --encrypt --sign %?a?-u %a? --armor --always-trust -- -r %r -- %f"
+#
+set pgp_encrypt_sign_command="/usr/lib/mutt/pgpewrap gpg %?p?--pinentry-mode loopback --passphrase-fd 0? --batch --quiet --no-verbose --textmode --output - %?a?--local-user %a? --armor -- --recipient %r -- --sign --encrypt %f"
```
please keep `/usr/lib/neomutt/`

gahr commented on contrib/gpg.rc:
```diff
@@ -29,58 +35,72 @@
-
-# set pgp_good_sign="^gpgv?: Good signature from "
-
-# OK, here's a version which uses gnupg's message catalog:
```
This command was probably useful.

libBletchley replied:
I was merging a couple versions, and didn't notice the `v?` difference from this one:
`# set pgp_good_sign="^gpg: Good signature from"`

Indeed the `v?` is good to have.

gahr replied:
I meant, this comment: `-# OK, here's a version which uses gnupg's message catalog:`

libBletchley commented:
Is there anything left for me to do?  Do you want the comment anchors removed?

flatcap commented:
That looks good @libBletchley

Before I merge it, please can you:
- Squash the commits into one neatly described commit
- Fix your commit credentials.  The commits currently say "GIT user"
- Send @flatcap an email, so he can send you an project invitation

Thanks
